File: JavaToKotlin/app/src/main/java/io/dustin/apps/board/api/usecase/user/WriteUserUseCase.java
package io.dustin.apps.board.api.usecase.user;

import io.dustin.apps.board.domain.user.model.SiteUser;
import io.dustin.apps.board.domain.user.service.ReadUserService;
import io.dustin.apps.board.domain.user.service.WriteUserService;
import io.dustin.apps.board.domain.user.model.dto.UserDto;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class WriteUserUseCase {

    private final ReadUserService readUserService;
    private final WriteUserService writeUserService;

    public UserDto execute(String nickName, String email, String password) {
        SiteUser siteUser = writeUserService.create(nickName,email,password);
        UserDto dto = UserDto.from(siteUser);
        return UserDto.from(siteUser);
    }

    public SiteUser getUser(Long userID){
        /**
         * null이 나올 수 있는 데이터도 dto로 줘야 하나 ??
          */
        SiteUser siteUser = readUserService.getUser(userID);
        return siteUser;

    }



}

File: dustin/src/main/kotlin/io/dustin/api/usercase/user/WriteUserUseCase.kt
package io.dustin.api.usercase.user

import io.dustin.api.usercase.user.model.CreateUser
import io.dustin.api.usercase.user.model.UpdateUser
import io.dustin.common.exception.BadParameterException
import io.dustin.domain.user.model.User
import io.dustin.domain.user.service.ReadUserService
import io.dustin.domain.user.service.WriteUserService
import org.springframework.stereotype.Service
import reactor.core.publisher.Mono

@Service
class WriteUserUseCase(
    private val read: ReadUserService,
    private val write: WriteUserService,
) {

    fun insert(command: CreateUser): Mono<User> {
        val created = User(name = command.name, job = command.job)
        return write.create(created)
    }

    /**
     * flatMap을 사용하지 않는다면 이 결과는 Mono<Mono<Musician>>이라는 요상한 형식이 된다.
     */
    fun update(id: Long, command: UpdateUser): Mono<User> {
        return read.userByIdOrThrow(id).flatMap { user ->
            val (user, assignments) = command.createAssignments(user)
            write.update(user, assignments)
        }.onErrorResume {
            Mono.error(BadParameterException(it.message))
        }
            .then(read.userById(id))
    }

}